File: frontend/src/lib/api/cooperative.ts
// frontend/src/lib/api/cooperative.ts
// Deliberately separate from firmChat.ts per spec section 3 isolation requirement.
import api from '@/lib/api'

export interface CooperativeMessage {
  id: string
  room_id: string
  author_handle: string
  body: string
  created_at: string
}

export interface CooperativeRoom {
  id: string
  room_type: string
  name: string | null
}

export const cooperativeApi = {
  optIn: async (): Promise<{ cooperative_enabled: boolean; handle: string }> => {
    const { data } = await api.post('/cooperative/opt-in')
    return data as { cooperative_enabled: boolean; handle: string }
  },

  getRooms: async (): Promise<{ items: CooperativeRoom[]; total: number }> => {
    const { data } = await api.get('/cooperative/rooms')
    return data as { items: CooperativeRoom[]; total: number }
  },

  getMessages: async (roomId: string, page = 1): Promise<{ items: CooperativeMessage[]; total: number }> => {
    const { data } = await api.get(`/cooperative/rooms/${roomId}/messages`, {
      params: { page, page_size: 100 },
    })
    return data as { items: CooperativeMessage[]; total: number }
  },

  postMessage: async (roomId: string, body: string): Promise<CooperativeMessage> => {
    const { data } = await api.post(`/cooperative/rooms/${roomId}/messages`, { body })
    return data as CooperativeMessage
  },
}

File: app/models/cooperative.py
# app/models/cooperative.py
#
# Deliberately separate from app/models/firm_chat.py per spec section 3:
# no shared models, no shared query helpers with any firm-scoped surface.
# Message authorship resolves only through CooperativeMember, never users.id directly.

import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, ForeignKey, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base_class import Base


class CooperativeMember(Base):
    """One record per user who has been granted access to the Growth Cooperative."""

    __tablename__ = "cooperative_members"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)

    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
        index=True,
    )

    firm_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("firms.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    handle: Mapped[str] = mapped_column(
        String(64),
        nullable=False,
        unique=True,
    )

    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    is_jamm_team: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    granted_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )


class CooperativeRoom(Base):
    """A room in the Growth Cooperative. Starts with a singleton "main" room."""

    __tablename__ = "cooperative_rooms"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)

    room_type: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
    )

    name: Mapped[str | None] = mapped_column(
        String(128),
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )


class CooperativeMessage(Base):
    """A message posted in a CooperativeRoom.

    author_member_id references CooperativeMember, never users.id directly,
    so a deactivated member's past messages survive intact.
    """

    __tablename__ = "cooperative_messages"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)

    room_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("cooperative_rooms.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    author_member_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("cooperative_members.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    body: Mapped[str] = mapped_column(Text, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
